pkg/reconciler/kubernetes/tektonhub/tektonhub.go

```diff
@@ -0,0 +1,595 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tektonhub
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	mf "github.com/manifestival/manifestival"
+	"github.com/tektoncd/operator/pkg/apis/operator/v1alpha1"
+	clientset "github.com/tektoncd/operator/pkg/client/clientset/versioned"
+	tektonhubconciler "github.com/tektoncd/operator/pkg/client/injection/reconciler/operator/v1alpha1/tektonhub"
+	"github.com/tektoncd/operator/pkg/reconciler/common"
+	"k8s.io/client-go/kubernetes"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/logging"
+	pkgreconciler "knative.dev/pkg/reconciler"
+)
+
+// Reconciler implements controller.Reconciler for TektonHub resources.
+type Reconciler struct {
+	// kubeClientSet allows us to talk to the k8s for core APIs
+	kubeClientSet kubernetes.Interface
+	// operatorClientSet allows us to configure operator objects
+	operatorClientSet clientset.Interface
+	// manifest is empty, but with a valid client and logger. all
+	// manifests are immutable, and any created during reconcile are
+	// expected to be appended to this one, obviating the passing of
+	// client & logger
+	manifest mf.Manifest
+	// Platform-specific behavior to affect the transform
+	extension common.Extension
+	// enqueueAfter enqueues a obj after a duration
+	enqueueAfter func(obj interface{}, after time.Duration)
+}
+
+var (
+	apiConfigMapName string = "api"
+	errKeyMissing    error  = fmt.Errorf("secret doesn't contains all the keys")
+	namespace        string = ""
+	// Check that our Reconciler implements controller.Reconciler
+	_ tektonhubconciler.Interface = (*Reconciler)(nil)
+	_ tektonhubconciler.Finalizer = (*Reconciler)(nil)
+)
+
+const (
+	dbInstallerSet          = "DbInstallerSet"
+	dbMigrationInstallerSet = "DbMigrationInstallerSet"
+	apiInstallerSet         = "ApiInstallerSet"
+
+	releaseVersionKey  = "operator.tekton.dev/release-version"
+	createdByKey       = "operator.tekton.dev/created-by"
+	targetNamespaceKey = "operator.tekton.dev/target-namespace"
+	createdByValue     = "TektonHub"
+)
+
+// FinalizeKind removes all resources after deletion of a TektonHub.
+func (r *Reconciler) FinalizeKind(ctx context.Context, original *v1alpha1.TektonHub) pkgreconciler.Event {
+	logger := logging.FromContext(ctx)
+
+	installerSets := original.Status.HubInstallerSet
+	if len(installerSets) == 0 {
+		return nil
+	}
+
+	for _, value := range installerSets {
+		err := r.operatorClientSet.OperatorV1alpha1().TektonInstallerSets().Delete(ctx, value, metav1.DeleteOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
```

Comment:
This should be handled using a List query. We don't use name(s) to query installerSets anymore. It can cause rogue un-garbage collected installerSets. 

We need to set labels (created-by, type and verision). then in finalize use the `created-by` and may be `type` label to make a List query and delete all that is returned

examples:
- https://github.com/tektoncd/operator/blob/a0cd67e5c428babf1f76316a14cf9c4526055cab/pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go#L100
- https://github.com/tektoncd/operator/blob/a0cd67e5c428babf1f76316a14cf9c4526055cab/pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go#L360